File: lld/problems/ride-sharing/go/ride_sharing.go
// Package ridesharing implements the classic Uber-style Ride Sharing LLD problem:
// rider/driver matching (Strategy), trip lifecycle, and pluggable pricing.
package ridesharing

import (
	"errors"
	"fmt"
	"math"
	"sync"
	"time"
)

type Location struct {
	X, Y float64
}

func distance(a, b Location) float64 {
	dx := a.X - b.X
	dy := a.Y - b.Y
	return math.Sqrt(dx*dx + dy*dy)
}

type Rider struct {
	ID   string
	Name string
}

type Driver struct {
	ID        string
	Name      string
	Location  Location
	available bool
}

type TripStatus int

const (
	Requested TripStatus = iota
	Accepted
	InProgress
	Completed
	Cancelled
)

type Trip struct {
	ID          string
	Rider       *Rider
	Driver      *Driver
	Pickup      Location
	Dropoff     Location
	Status      TripStatus
	RequestedAt time.Time
	Fare        float64
}

// DriverMatchingStrategy picks a driver for a trip from the available pool.
type DriverMatchingStrategy interface {
	Match(pickup Location, drivers []*Driver) *Driver
}

// NearestAvailableDriverStrategy picks the closest available driver to the pickup point.
type NearestAvailableDriverStrategy struct{}

func (NearestAvailableDriverStrategy) Match(pickup Location, drivers []*Driver) *Driver {
	var best *Driver
	bestDist := math.MaxFloat64
	for _, d := range drivers {
		if !d.available {
			continue
		}
		dist := distance(pickup, d.Location)
		if dist < bestDist {
			bestDist = dist
			best = d
		}
	}
	return best
}

// PricingStrategy computes the fare for a completed trip.
type PricingStrategy interface {
	CalculateFare(t *Trip) float64
}

// DistanceBasedPricing charges a base fare plus a per-unit-distance rate.
type DistanceBasedPricing struct {
	BaseFare    float64
	PerDistance float64
}

func (p DistanceBasedPricing) CalculateFare(t *Trip) float64 {
	return p.BaseFare + p.PerDistance*distance(t.Pickup, t.Dropoff)
}

var (
	ErrTripNotFound      = errors.New("trip not recognized")
	ErrNoDriverAvailable = errors.New("no driver available")
	ErrInvalidTransition = errors.New("invalid trip status transition")
)

type RideSharingSystem struct {
	Matching DriverMatchingStrategy
	Pricing  PricingStrategy

	mu      sync.Mutex
	drivers []*Driver
	trips   map[string]*Trip
	seq     int
}

func NewRideSharingSystem(drivers []*Driver, matching DriverMatchingStrategy, pricing PricingStrategy) *RideSharingSystem {
	return &RideSharingSystem{
		Matching: matching,
		Pricing:  pricing,
		drivers:  drivers,
		trips:    make(map[string]*Trip),
	}
}

func (s *RideSharingSystem) RequestRide(rider *Rider, pickup, dropoff Location) *Trip {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.seq++
	trip := &Trip{
		ID:          fmt.Sprintf("TR-%d", s.seq),
		Rider:       rider,
		Pickup:      pickup,
		Dropoff:     dropoff,
		Status:      Requested,
		RequestedAt: time.Now(),
	}
	s.trips[trip.ID] = trip
	return trip
}

// MatchDriver finds the nearest available driver and assigns it atomically:
// the whole find-and-mark-unavailable step is guarded by the system mutex so
// two concurrent calls can never both be assigned the same single driver.
func (s *RideSharingSystem) MatchDriver(tripID string) (*Driver, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	trip, ok := s.trips[tripID]
	if !ok {
		return nil, ErrTripNotFound
	}
	if trip.Status != Requested {
		return nil, ErrInvalidTransition
	}

	driver := s.Matching.Match(trip.Pickup, s.drivers)
	if driver == nil {
		return nil, ErrNoDriverAvailable
	}

	driver.available = false
	trip.Driver = driver
	trip.Status = Accepted
	return driver, nil
}

func (s *RideSharingSystem) StartTrip(tripID string) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	trip, ok := s.trips[tripID]
	if !ok {
		return ErrTripNotFound
	}
	if trip.Status != Accepted {
		return ErrInvalidTransition
	}
	trip.Status = InProgress
	return nil
}

func (s *RideSharingSystem) CompleteTrip(tripID string) (float64, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	trip, ok := s.trips[tripID]
	if !ok {
		return 0, ErrTripNotFound
	}
	if trip.Status != InProgress {
		return 0, ErrInvalidTransition
	}
	trip.Status = Completed
	trip.Fare = s.Pricing.CalculateFare(trip)
	if trip.Driver != nil {
		trip.Driver.available = true
	}
	return trip.Fare, nil
}

func (s *RideSharingSystem) CancelTrip(tripID string) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	trip, ok := s.trips[tripID]
	if !ok {
		return ErrTripNotFound
	}
	if trip.Status != Requested && trip.Status != Accepted {
		return ErrInvalidTransition
	}
	trip.Status = Cancelled
	if trip.Driver != nil {
		trip.Driver.available = true
	}
	return nil
}

File: lld/problems/ride-sharing/python/ride_sharing.py
"""Ride Sharing LLD — Python reference implementation.

Rider/driver matching (Strategy), trip lifecycle (state machine), and
pluggable pricing. See ../README.md for the design writeup.
"""
from __future__ import annotations

import math
import threading
from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum, auto
from typing import List, Optional


@dataclass(frozen=True)
class Location:
    x: float
    y: float

    def distance_to(self, other: "Location") -> float:
        return math.sqrt((self.x - other.x) ** 2 + (self.y - other.y) ** 2)


@dataclass
class Rider:
    id: str
    name: str


@dataclass
class Driver:
    id: str
    name: str
    location: Location
    available: bool = True


class TripStatus(Enum):
    REQUESTED = auto()
    ACCEPTED = auto()
    IN_PROGRESS = auto()
    COMPLETED = auto()
    CANCELLED = auto()


@dataclass
class Trip:
    id: str
    rider: Rider
    pickup: Location
    dropoff: Location
    driver: Optional[Driver] = None
    status: TripStatus = TripStatus.REQUESTED
    requested_at: datetime = field(default_factory=datetime.now)
    fare: float = 0.0


class TripNotFoundError(Exception):
    """Raised when a trip ID is not recognized."""


class NoDriverAvailableError(Exception):
    """Raised when no available driver can be matched."""


class InvalidTransitionError(Exception):
    """Raised when a trip status transition is not legal."""


class DriverMatchingStrategy(ABC):
    """Picks a driver for a trip from the available pool."""

    @abstractmethod
    def match(self, pickup: Location, drivers: List[Driver]) -> Optional[Driver]:
        ...


class NearestAvailableDriverStrategy(DriverMatchingStrategy):
    """Picks the closest available driver to the pickup point."""

    def match(self, pickup: Location, drivers: List[Driver]) -> Optional[Driver]:
        best: Optional[Driver] = None
        best_dist = math.inf
        for d in drivers:
            if not d.available:
                continue
            dist = pickup.distance_to(d.location)
            if dist < best_dist:
                best_dist = dist
                best = d
        return best


class PricingStrategy(ABC):
    """Computes the fare for a completed trip."""

    @abstractmethod
    def calculate_fare(self, trip: Trip) -> float:
        ...


@dataclass
class DistanceBasedPricing(PricingStrategy):
    """Charges a base fare plus a per-unit-distance rate."""

    base_fare: float
    per_distance: float

    def calculate_fare(self, trip: Trip) -> float:
        return self.base_fare + self.per_distance * trip.pickup.distance_to(trip.dropoff)


class RideSharingSystem:
    """Facade coordinating the driver pool, trip map, matching, and pricing.

    All lifecycle methods acquire a single system lock so that, in
    particular, the find-and-mark-unavailable step in match_driver is
    atomic: two concurrent calls can never both be assigned the same
    single driver.
    """

    def __init__(self, drivers: List[Driver], matching: DriverMatchingStrategy, pricing: PricingStrategy):
        self.matching = matching
        self.pricing = pricing
        self._lock = threading.Lock()
        self._drivers = drivers
        self._trips: dict[str, Trip] = {}
        self._seq = 0

    def request_ride(self, rider: Rider, pickup: Location, dropoff: Location) -> Trip:
        with self._lock:
            self._seq += 1
            trip = Trip(id=f"TR-{self._seq}", rider=rider, pickup=pickup, dropoff=dropoff)
            self._trips[trip.id] = trip
            return trip

    def match_driver(self, trip_id: str) -> Driver:
        """Finds the nearest available driver and assigns it atomically."""
        with self._lock:
            trip = self._trips.get(trip_id)
            if trip is None:
                raise TripNotFoundError(trip_id)
            if trip.status != TripStatus.REQUESTED:
                raise InvalidTransitionError(trip.status)

            driver = self.matching.match(trip.pickup, self._drivers)
            if driver is None:
                raise NoDriverAvailableError()

            driver.available = False
            trip.driver = driver
            trip.status = TripStatus.ACCEPTED
            return driver

    def start_trip(self, trip_id: str) -> None:
        with self._lock:
            trip = self._trips.get(trip_id)
            if trip is None:
                raise TripNotFoundError(trip_id)
            if trip.status != TripStatus.ACCEPTED:
                raise InvalidTransitionError(trip.status)
            trip.status = TripStatus.IN_PROGRESS

    def complete_trip(self, trip_id: str) -> float:
        with self._lock:
            trip = self._trips.get(trip_id)
            if trip is None:
                raise TripNotFoundError(trip_id)
            if trip.status != TripStatus.IN_PROGRESS:
                raise InvalidTransitionError(trip.status)
            trip.status = TripStatus.COMPLETED
            trip.fare = self.pricing.calculate_fare(trip)
            if trip.driver is not None:
                trip.driver.available = True
            return trip.fare

    def cancel_trip(self, trip_id: str) -> None:
        with self._lock:
            trip = self._trips.get(trip_id)
            if trip is None:
                raise TripNotFoundError(trip_id)
            if trip.status not in (TripStatus.REQUESTED, TripStatus.ACCEPTED):
                raise InvalidTransitionError(trip.status)
            trip.status = TripStatus.CANCELLED
            if trip.driver is not None:
                trip.driver.available = True


def _demo() -> None:
    drivers = [
        Driver("D1", "Alice", Location(0, 0)),
        Driver("D2", "Bob", Location(10, 10)),
    ]
    system = RideSharingSystem(drivers, NearestAvailableDriverStrategy(), DistanceBasedPricing(2, 1.5))

    rider = Rider("R1", "Riya")
    trip = system.request_ride(rider, Location(0, 0), Location(3, 4))
    print(f"Requested trip {trip.id} status={trip.status}")

    driver = system.match_driver(trip.id)
    print(f"Matched driver {driver.id} status={trip.status}")

    system.start_trip(trip.id)
    print(f"Started trip, status={trip.status}")

    fare = system.complete_trip(trip.id)
    print(f"Completed trip, fare={fare} status={trip.status}")


if __name__ == "__main__":
    _demo()
